fv-tests/server_test.go

```diff
@@ -0,0 +1,710 @@
+		loop:
+			for {
+				select {
+				case <-timeout.C:
+					break loop
+				case <-oneSec.C:
+					// Check the rate is in the right ballpark: after one second we should have
+					// dropped approximately 20 clients.
+					Expect(numFinished).To(BeNumerically(">", 10))
+					Expect(numFinished).To(BeNumerically("<", 30))
+				case c := <-finishedC:
+					numFinished += c
+				}
+			}
+			// After the timeout we should have dropped exactly the right number of connections.
+			Expect(numFinished).To(Equal(40))
+			expectGaugeValue("typha_connections_active", 60.0)
+		})
+
+		It("should pass through a KV and status", func() {
+			decoupler.OnStatusUpdated(api.ResyncInProgress)
+			decoupler.OnUpdates([]api.Update{configFoobarBazzBiff})
+			decoupler.OnStatusUpdated(api.InSync)
+			expectClientStates(
+				api.InSync,
+				map[string]api.Update{
+					"/calico/v1/config/foobar": configFoobarBazzBiff,
+				},
+			)
+		})
+
+		It("should pass through many KVs", func() {
+			expectedEndState := sendNUpdatesThenInSync(1000)
+			expectClientStates(api.InSync, expectedEndState)
+		})
+
+		It("should report the correct number of connections", func() {
+			expectGaugeValue("typha_connections_active", 100.0)
+		})
+
+		It("should report the correct number of connections after killing the clients", func() {
+			for _, c := range clientStates {
+				c.clientCancel()
+			}
+			expectGaugeValue("typha_connections_active", 0.0)
+		})
+
+		It("with churn, it should report the correct number of connections after killing the clients", func() {
+			// Generate some churn while we disconnect the clients.
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func() {
+				sendNUpdatesThenInSync(1000)
+				wg.Done()
+			}()
+			defer wg.Wait()
+			for _, c := range clientStates {
+				c.clientCancel()
+				time.Sleep(100 * time.Microsecond)
+			}
+			expectGaugeValue("typha_connections_active", 0.0)
+		})
+	})
+})
+
+var _ = Describe("With an in-process Server with short ping timeout", func() {
+	var cacheCxt context.Context
+	var cacheCancel context.CancelFunc
+	var cache *snapcache.Cache
+	var server *syncserver.Server
+	var serverCxt context.Context
+	var serverCancel context.CancelFunc
+	var serverAddr string
+
+	BeforeEach(func() {
+		cache = snapcache.New(snapcache.Config{
+			// Set the batch size small so we can force new Breadcrumbs easily.
+			MaxBatchSize: 10,
+			// Reduce the wake up interval from the default to give us faster tear down.
+			WakeUpInterval: 50 * time.Millisecond,
+		})
+		server = syncserver.New(cache, syncserver.Config{
+			PingInterval: 100 * time.Millisecond,
+			PongTimeout:  500 * time.Millisecond,
+			Port:         syncserver.PortRandom,
+			DropInterval: 50 * time.Millisecond,
+		})
+		cacheCxt, cacheCancel = context.WithCancel(context.Background())
+		cache.Start(cacheCxt)
+		serverCxt, serverCancel = context.WithCancel(context.Background())
+		server.Start(serverCxt)
+		serverAddr = fmt.Sprintf("127.0.0.1:%d", server.Port())
```

Comment:
It's a bit awkward to do that, means returning a lot of state or introducing more ginkgo nesting, think I'd marginally prefer to live-with.